### PR TITLE
Heap render debugging facility

### PR DESCRIPTION
**Pull request by nwf**
This has been kicking around locally and undergoing the occasional modification, but it might be useful to push into the tree even if we never use it in most cases?

### CODE REVIEW

**Review by davidchisnall — COMMENTED** [review]
Looks useful, but shouldn't be behind `NDEBUG`.

**Review by rmn30 — APPROVED** [review]
I haven't reviewed the render code in detail but given this is entirely guarded by ifdefs I can't see any harm in merging. Maybe add a call to `render_heap` in a test?

**Comment by rmn30** [comment]
I tried this locally by adding a call to `heap_render` in the allocator test. It seems to work but I had to pass both `--debug-allocator=y` and `--allocator-rendering=y` to `xmake config` to get it to show anything. We should add this call in the test and enable the flag in CI debug builds.

**Comment by nwf** [comment]
Hm.  Maybe we shouldn't use the generic `Debug` so that `--allocator-rendering` and `--debug-allocator` are orthogonal.  Have pushed patch to that effect; please advise.

**Comment by rmn30** [comment]
That's reasonable but I wonder if we also want a mode that will just do the assertions without the verbosity? Probably something we should add if we ever want it.

**Comment by nwf** [comment]
> That's reasonable but I wonder if we also want a mode that will just do the assertions without the verbosity? Probably something we should add if we ever want it.

Oh, that's a great idea and easy, just move the `using RenderDebug` into the function and add another template paramater... done. :)

**Comment by rmn30** [comment]
I don't think that will work as written because `ConditionalDebug::Assert` is only enabled when the condition is true. I think it will work if you use `ConditionalDebug::Invariant` instead.

**Comment by nwf** [comment]
Some *truly* terrifying C++ later...

File: sdk/core/allocator/alloc.h [hunk 1]
```diff
@@ -2811,4 +2811,163 @@ class MState
 	{
 		ABORT();
 	}
+
+#if HEAP_RENDER
+	public:
+	/**
+	 * "Render" the heap for debugging.
+	 */
+	template<bool Asserts = false, bool Chatty = true>
```
Comment (rmn30) [inline sdk/core/allocator/alloc.h:2820]: I think Asserts should default to `true`. They don't look that expensive compared to the debug logs (I guess `ds::linked_list::is_well_formed` may be linear in length of list?).

**Review by rmn30 — APPROVED** [review]
Looks fine. I will defer to @davidchisnall  for the `debug.hh` bits...

File: sdk/core/allocator/alloc.h (hunk 1 above)
Comment (nwf) [reply sdk/core/allocator/alloc.h]: Thanks!

Comment (nwf) [reply sdk/core/allocator/alloc.h]: > I guess `ds::linked_list::is_well_formed` may be linear in length of list?

No, is just a check of the well-formedness of the node.  Could probably be better named, in retrospect.

**Pull request merged**